var/spack/repos/builtin/packages/hypre/package.py

```diff
@@ -35,12 +35,13 @@ class Hypre(Package):
     homepage = "http://computation.llnl.gov/project/linear_solvers/software.php"
     url      = "http://computation.llnl.gov/project/linear_solvers/download/hypre-2.10.0b.tar.gz"
 
-    version('develop', git='https://github.com/LLNL/hypre', tag='master')
-    version('xsdk-0.2.0', git='https://github.com/LLNL/hypre', tag='xsdk-0.2.0')
+    version('2.12.1', 'c6fcb6d7e57cec1c7ce4a44da885068c', url='https://github.com/LLNL/hypre/archive/v2.12.1.tar.gz')
     version('2.11.2', 'd507943a1a3ce5681c3308e2f3a6dd34')
     version('2.11.1', '3f02ef8fd679239a6723f60b7f796519')
     version('2.10.1', 'dc048c4cabb3cd549af72591474ad674')
     version('2.10.0b', '768be38793a35bb5d055905b271f5b8e')
+    version('develop', git='https://github.com/LLNL/hypre', tag='master')
```

Comment:
honestly, I think it should be `branch='master'` instead of `tag='master'`.

Reply:
i did not change it in this PR, only move those at the bottom

Reply:
but i agree, `branch=master` is probably what it should be

Reply:
Thank you for the help.  Being new to Spack, though, I have a few questions.

- The xsdk-0.2.0 version of hypre was removed.  Does this break compatibility between the hypre package and xsdk-0.2.0?

- Why does the xSDK literature tell me to list the current hypre release (here v2.12.1) for the xSDK as "preferred=True"?  This was included in the original pull request #5643, but removed by davydden here in this pull request.  Maybe the xSDK literature needs to be updated?  

- Is the below command invalid or deprecated with regard to the command davydden used above?
  
  version('2.12.1', git='https://github.com/LLNL/hypre', tag='v2.12.1')
  
  It's short and easy to read, so I'm a little confused.

Todd: you're maybe a good bridge here (being a member of both xSDK and Spack) to answer these questions. @tgamblin

Reply:
`preferred=True` isn't needed as the newest version is always preferred. And `version()` with a checksum is preferred over the one with a tag.

Reply:
OK.  Any knowledge of how the removal of the xsdk-0.2.0 version affects the xSDK?  Or should Todd answer that?  

I left a comment in the xSDK literature about not using "preferred=True", when the version being marked is the newest version.

Reply:
@jbschroder:

> The xsdk-0.2.0 version of hypre was removed. Does this break compatibility between the hypre package and xsdk-0.2.0?

It was just moved lower -- see @davydden's comment.

> Why does the xSDK literature tell me to list the current hypre release (here v2.12.1) for the xSDK as "preferred=True"? This was included in the original pull request #5643, but removed by davydden here in this pull request. Maybe the xSDK literature needs to be updated?

I put that in the xSDK documentation to be explicit; Spack will automatically pick the newest, non-develop version if nothing is marked `preferred`.  @junghans is technically right -- it's not needed given Spack's current version selection policy.  I think it's up to xSDK folks whether they want to put it there.

One nit: If you have both `3.2.1` and `3.2.1rc1`, Spack will currently think that 3.2.1rc1 is newer.  So you may want to be explicit if you're using `rc` tags for release candidates.  The other option would be to remove the `rc` versions when you add the final release.

> Is the below command invalid or deprecated with regard to the command davydden used above?
> 
> version('2.12.1', git='https://github.com/LLNL/hypre', tag='v2.12.1')
> 
> It's short and easy to read, so I'm a little confused.

@junghans is right that we prefer versions with checksums over versions with tags.  That might be worth adding to the docs as well.  The main reason to prefer the checksum over the tag is that it's more of a guarantee that what the user downloads is what the packager intended.  Tags can move.

Reply:
Thanks Todd!  That makes sense.  I added a comment to the xSDK workflow document about the use of the preferred=True parameter.